### PR TITLE
Adding `Period_prior_to_index` function

### PR DESCRIPTION
**Pull request by Jay-sanjay**
**This PR is related to #1**

## PR CheckList

- [x] Added the very first dispatch for `period_prior_to_index` function
- [x] Added the test-suite for the same
- [x]  Added the Doc-strings  
- [x]  Added the remaining dispatches with test-suite

### CODE REVIEW

File: src/preprocessing.jl [hunk 1]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+end
+
+
+
+
+
+
+
+
+
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Could you remove this now?

File: src/preprocessing.jl [hunk 2]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+"""
+function Period_prior_to_index(cohort_id::Vector, conn; date_prior=Day(100), tab=cohort)
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Could you make this lowercase? i.e. `period_prior_to_index` ?

File: src/preprocessing.jl [hunk 3]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+
+Given a vector of cohort IDs, this function returna a DataFrame with the cohort_start_date adjusted by the date_prior.
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Could we rephrase this to the following:

"Given `cohort_id`'s, return a `DataFrame` with the `cohort_start_date` adjusted to prior each subjects' cohort entry date (i.e. their `cohort_start_date`).

File: src/preprocessing.jl [hunk 4]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+
+# Arguments:
+
+- `cohort_id` - vector of cohort IDs
+- `conn` - database connection
+
+# Keyword Arguments:
+
+- `date_prior` - period prior to the index date; default is 100 days
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Could we update this as:

`date_prior::Dates.AbstractTime` - how much time prior the index date should be adjusted by; accepts a `Dates.AbstractTime` object such as `Day`, `Month`, etc. (Default: `Day(100)`)

File: src/preprocessing.jl [hunk 5]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+- `tab` - the `SQLTable` representing the cohort table; default `cohort`
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Small tweak but could you say:

- `tab` - the `SQLTable` representing the cohort table. (Default: `cohort`)

File: src/preprocessing.jl [hunk 6]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+
+# Returns
+
+- DataFrame with the cohort_start_date adjusted by the date_prior.
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Small tweak:

`cohort_start_date` adjusted by the `date_prior`

File: src/preprocessing.jl [hunk 7]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+    sql = From(tab) |>
+            Where(Fun.in(Get.cohort_definition_id, cohort_id...)) |>
+            Select(Get.cohort_definition_id, Get.subject_id, Get.cohort_start_date) |>
+            q -> render(q, dialect=dialect)
+
+    # Execute the SQL query and fetch the result into a DataFrame
+    df = DBInterface.execute(conn, String(sql)) |> DataFrame
+
+    # Check if the DataFrame is not empty
+    # TODO: Is it really necessary to add this check ??
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: We concluded this would be a nice check to see if the dataframe is not empty. So you can get rid of this comment.

File: test/Data-Preprocessing/preprocessing.jl [hunk 8]
```diff
@@ -6,8 +6,50 @@ using Test
     test_drug_exposure_start_date = [-3.727296e8, 2.90304e7, -5.333472e8, -8.18208e7]
     test_df1 = DataFrame(drug_exposure_id = test_drug_exposure_ids, drug_exposure_start_date = test_drug_exposure_start_date)
     result =  Dummy(test_drug_exposure_ids, sqlite_conn)
-
+    println(result)
```
Comment (TheCedarPrince) [inline test/Data-Preprocessing/preprocessing.jl]: Could you get rid of this line? And do you still need the dummy test or can they be removed?

File: src/preprocessing.jl [hunk 9]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+
+# Arguments:
+
+- `person_ids` - vector of person IDs
+- `start_date_on_person` - function that returns the SQL query to get the start date of the person
+- `conn` - database connection
+
+# Keyword Arguments:
+
+- `date_prior` - period prior to the index date; default is 100 days
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Since this method only accepts a function that modifies the `cohort_start_date` per person, this argument and its usage should be deleted.

File: src/preprocessing.jl [hunk 10]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+
+"""
+function Period_prior_to_index(person_ids::Vector, start_date_on_person::Function, conn; date_prior=Day(100))
+
+Given a vector of person IDs, this function returns a DataFrame with the cohort_start_date adjusted by the date_prior.
+
+# Arguments:
+
+- `person_ids` - vector of person IDs
+- `start_date_on_person` - function that returns the SQL query to get the start date of the person
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: I would suggest changing this name to `index_date_func` and in the docstring specifies that the function should accept three arguments being `df` (the dataframe computed for the cohorts),  `tables`, and `conn` with a note that it should return a DataFrame and not SQL. I think the function a user passes in should allow for execution of SQL and additional post-processing if they need it.

P.S. This will be cool as that means it could accept TidierDB queries someday down the road too.

File: src/preprocessing.jl [hunk 11]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+        start_date_on_person::Function, 
+        conn; 
+        date_prior=Day(100)
+    )
+
+    tables = GenerateTables(conn, inplace = false, exported=true)
+
+    sql = start_date_on_person(person_ids, tables)
+
+    df = DBInterface.execute(conn, String(sql)) |> DataFrame
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: You'll need to put this line before sql to compute the cohorts.

File: src/preprocessing.jl [hunk 12]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+# Keyword Arguments:
+
+- `date_prior` - period prior to the index date; default is 100 days
+
+# Returns
+
+- DataFrame with the cohort_start_date adjusted by the date_prior.
+"""
+function Period_prior_to_index(
+        person_ids::Vector,
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Shouldn't this be cohorts?

File: src/preprocessing.jl [hunk 13]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+    )
+
+    tables = GenerateTables(conn, inplace = false, exported=true)
+
+    sql = start_date_on_person(person_ids, tables)
+
+    df = DBInterface.execute(conn, String(sql)) |> DataFrame
+
+    # Check if the DataFrame is not empty
+    # TODO: Is it really necessary to add this check ??
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: I think this section should be removed since you no longer have `date_prior`

File: test/Data-Preprocessing/preprocessing.jl [hunk 14]
```diff
@@ -6,8 +6,50 @@ using Test
...
     @test test_drug_exposure_start_date == result.drug_exposure_start_date[1:4]
     @test test_drug_exposure_ids == result.drug_exposure_id[1:4]
+    
+    MakeTables(sqlite_conn, :sqlite, "main")
+    test_cohort_ids = [1.0, 1.0, 1.0, 1.0]
+    # test_df2 = DataFrame(cohort_definition_id = test_cohort_definition_id, cohort_id = test_cohort_ids)
+    df = Dummy2(test_cohort_ids, sqlite_conn)
+    println(first(df, 5))
```
Comment (TheCedarPrince) [inline test/Data-Preprocessing/preprocessing.jl]: Should these tests be cleaned up here?

File: test/Data-Preprocessing/preprocessing.jl [hunk 15]
```diff
@@ -6,8 +6,50 @@ using Test
...
+        Select(Get.cohort_definition_id, Get.subject_id, Get.cohort_start_date) |>
+        q -> render(q, dialect = :sqlite)
+
+        return sql
+    end
+
+    result = Period_prior_to_index(test_person_ids, start_date_on_person, sqlite_conn)
+    
+    @test test_person_ids == result.cohort_definition_id[1:5]
+    @test test_subject_ids == result.subject_id[1:5]
```
Comment (TheCedarPrince) [inline test/Data-Preprocessing/preprocessing.jl:43]: Can you make sure to clean the tests here? They seem fine but there might be a few extraneous prints here and there.

File: src/preprocessing.jl [hunk 16]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+
+# Keyword Arguments:
+
+- `date_prior` - period prior to the index date; default is 100 days
+- `tab` - the `SQLTable` representing the cohort table; default `cohort`
+
+# Returns
+
+- DataFrame with the cohort_start_date adjusted by the date_prior.
+"""
```
Comment (TheCedarPrince) [inline src/preprocessing.jl:131]: Could you add an example section here about how to call this function?

File: src/preprocessing.jl [hunk 17]
```diff
@@ -26,4 +26,122 @@ function Dummy(
...
+- `conn` - database connection
+
+# Keyword Arguments:
+
+- `date_prior` - period prior to the index date; default is 100 days
+
+# Returns
+
+- DataFrame with the cohort_start_date adjusted by the date_prior.
+"""
```
Comment (TheCedarPrince) [inline src/preprocessing.jl]: Could you add an example section here about how to call this function and how to pass in a function into the argument which accepts a user defined function? So show how to make a function? Thanks!

**Review by TheCedarPrince — CHANGES_REQUESTED** [review]
Hey @Jay-sanjay ,

This looks pretty good -- left some comments with the biggest suggested changes being in the second dispatch. Additionally, could you add an issue so we don't forget about it to later come back and add a dispatch for when you pass in just person IDs instead of cohort IDs? I am fine with us focusing on Cohort IDs first, but later, handling person IDs generally would be great too.

Great work so far Jay!

File: src/preprocessing.jl (hunk 1 above)
Comment (Jay-sanjay) [reply src/preprocessing.jl]: Sure !

File: test/Data-Preprocessing/preprocessing.jl (hunk 14 above)
Comment (Jay-sanjay) [reply test/Data-Preprocessing/preprocessing.jl]: Yes they can be removed now

File: src/preprocessing.jl (hunk 10 above)
Comment (Jay-sanjay) [reply src/preprocessing.jl]: @TheCedarPrince  but, then where do we adjust the dates based on `date-prior` ?

Comment (TheCedarPrince) [reply src/preprocessing.jl]: Oh so I imagine the user manually defines and applies the transformation to the start_date column within the function they pass in. Here's a simplified sketch with pseudo code:

```julia
my_func(df, tables, conn)

for patient in eachrow(df)
# Some criteria to meet
    if Criteria 1
        update patient.start_date by some change
# Some criteria to meet
    elseif Criteria 2

# ... Other additional create and transformations

return df

end
```

So the `date_prior` variable goes away and instead, the user passes in exactly how they want to transform each patient based on some criteria they are looking for in the function they provide. Does that make more sense?

**Comment by Jay-sanjay** [comment]
@TheCedarPrince  can we ignore the code-coverage for now ?

**Comment by TheCedarPrince** [comment]
@Jay-sanjay , if you could add one more test that involves checking the result of a defective DataFrame (like an empty dataframe), that would be perfect and get CodeCov back up to 100. Then we could merge this.

**Review by TheCedarPrince — APPROVED** [review]
Looks great to me! Let's merge. GREAT JOB!
